File: go/_2022/day3.go
package _2022

import (
	"advent/utils"
	"fmt"
	"strings"
	"unicode"
)

func getPriority(letter rune) int {
	if unicode.IsLower(letter) {
		return int(letter) - 97 + 1
	} else {
		return int(letter) - 65 + 26 + 1
	}
}

func getCommonLetter(words ...string) rune {
	first, rest := words[0], words[1:]
	for _, letter := range first {
		found := true
		for _, word := range rest {
			if !strings.ContainsAny(word, string(letter)) {
				found = false
				break // don't search next word; try next letter instead
			}
		}
		if found {
			return rune(letter)
		}
	}
	return 0
}

func Day3Part1() string {
	input := utils.LoadPuzzleInput("2022/day3")
	elves := strings.Split(input, "\n")
	score := 0
	for _, elf := range elves {
		middle := len(elf) / 2
		left, right := elf[:middle], elf[middle:]
		shared := getCommonLetter(left, right)
		score += getPriority(shared)
	}
	return fmt.Sprint(score)
}

func Day3Part2() string {
	input := utils.LoadPuzzleInput("2022/day3")
	elves := strings.Split(input, "\n")
	score := 0
	for ii := 0; ii < len(elves); ii += 3 {
		elf, second, third := elves[ii], elves[ii+1], elves[ii+2]
		shared := getCommonLetter(elf, second, third)
		score += getPriority(shared)
	}
	return fmt.Sprint(score)
}

func Day3() {
	utils.Profile(Day3Part1)
	utils.Profile(Day3Part2)
}

File: typescript/_2022/day3.ts
import {loadPuzzleInput, profile} from "../utils";


function getPriority(letter: string): number {
    if (letter.toLowerCase() === letter) {
        return letter.codePointAt(0) - 97 + 1
    } else {
        return letter.codePointAt(0) - 65 + 26 + 1
    }
}

function getCommonLetter(...words: string[]): string {
    let [first, ...rest] = words
    for (const letter of Array.from(first)) {
        let found = true
        for (const word of rest) {
            if (!word.includes(letter)) {
                found = false
                break
            }
        }
        if (found) {
            return letter
        }
    }
    return ""
}

export function day3Part1(): number {
    const input = loadPuzzleInput("2022/day3")
    const elves = input.split("\n")
    let score = 0
    for (const elf of elves) {
        const middle = elf.length / 2
        const [left, right] = [elf.slice(0, middle), elf.slice(middle, elf.length)]
        const shared = getCommonLetter(left, right)
        score += getPriority(shared)
    }
    return score
}

export function day3Part2(): number {
    const input = loadPuzzleInput("2022/day3")
    const elves = input.split("\n")
    let score = 0
    for (let ii = 0; ii < elves.length; ii += 3) {
        const [elf, second, third] = elves.slice(ii, ii + 3)
        const shared = getCommonLetter(elf, second, third)
        score += getPriority(shared)
    }
    return score
}

if (require.main === module) {
    profile(day3Part1)
    profile(day3Part2)
}